kvdb/postgres/readwrite_bucket.go

```diff
@@ -0,0 +1,436 @@
+// +build kvdb_postgres
+
+package postgres
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+
+	"github.com/btcsuite/btcwallet/walletdb"
+)
+
+// readWriteBucket stores the bucket id and the buckets transaction.
+type readWriteBucket struct {
+	// id is used to identify the bucket. If id is null, it refers to the
+	// root bucket.
+	id *int64
+
+	// tx holds the parent transaction.
+	tx *readWriteTx
+
+	table string
+}
+
+// newReadWriteBucket creates a new rw bucket with the passed transaction
+// and bucket id.
+func newReadWriteBucket(tx *readWriteTx, table string,
+	id *int64) *readWriteBucket {
+
+	return &readWriteBucket{
+		id:    id,
+		tx:    tx,
+		table: table,
+	}
+}
+
+// NestedReadBucket retrieves a nested read bucket with the given key.
+// Returns nil if the bucket does not exist.
+func (b *readWriteBucket) NestedReadBucket(key []byte) walletdb.ReadBucket {
+	return b.NestedReadWriteBucket(key)
+}
+
+func parentSelector(id *int64) string {
+	if id == nil {
+		return "parent_id IS NULL"
+	}
+	return fmt.Sprintf("parent_id=%v", *id)
+}
+
+// ForEach invokes the passed function with every key/value pair in
+// the bucket. This includes nested buckets, in which case the value
+// is nil, but it does not include the key/value pairs within those
+// nested buckets.
+func (b *readWriteBucket) ForEach(cb func(k, v []byte) error) error {
+	cursor := b.ReadWriteCursor()
+
+	k, v := cursor.First()
+	for k != nil {
+		err := cb(k, v)
+		if err != nil {
+			return err
+		}
+
+		k, v = cursor.Next()
+	}
+
+	return nil
+}
+
+// Get returns the value for the given key. Returns nil if the key does
+// not exist in this bucket.
+func (b *readWriteBucket) Get(key []byte) []byte {
+	// Return nil if the key is empty.
+	if len(key) == 0 {
+		return nil
+	}
+
+	var value *[]byte
+	err := b.tx.QueryRow(
+		"SELECT value FROM "+b.table+" WHERE "+parentSelector(b.id)+
+			" AND key=$1", key,
+	).Scan(&value)
+
+	switch {
+	case err == sql.ErrNoRows:
+		return nil
+
+	case err != nil:
+		panic(err)
```

Comment:
Why panic vs just return an error as normal?

Reply:
Unfortunately the `Get` method of the `kvdb` interface doesn't have an err return value.